custom_components/rpi_gpio/binary_sensor.py

```diff
@@ -1,141 +1,89 @@
+    hub = hass.data[DOMAIN]
+    if not hub._online:
+        _LOGGER.error("hub not online, bailing out")
 
-    ports = config[CONF_PORTS]
-    for port_num, port_name in ports.items():
+    sensors = []
+    for sensor in config.get(CONF_SENSORS):
         sensors.append(
-            RPiGPIOBinarySensor(
-                port_name, port_num, pull_mode, bouncetime, invert_logic
+            GPIODBinarySensor(
+                hub,
+                sensor[CONF_NAME],
+                sensor[CONF_PORT],
+                sensor.get(CONF_UNIQUE_ID) or f"{DOMAIN}_{sensor[CONF_PORT]}_{sensor[CONF_NAME].lower().replace(' ', '_')}",
+                sensor.get(CONF_INVERT_LOGIC),
+                sensor.get(CONF_PULL_MODE),
+                sensor.get(CONF_BOUNCETIME)
             )
         )
 
-    add_entities(sensors, True)
-
+    async_add_entities(sensors)
 
-class RPiGPIOBinarySensor(BinarySensorEntity):
-    """Represent a binary sensor that uses Raspberry Pi GPIO."""
 
-    async def async_read_gpio(self):
-        """Read state from GPIO."""
-        await asyncio.sleep(float(self._bouncetime) / 1000)
-        self._state = await self.hass.async_add_executor_job(read_input, self._port)
-        self.async_write_ha_state()
+class GPIODBinarySensor(BinarySensorEntity):
+    should_poll = False
 
-    def __init__(self, name, port, pull_mode, bouncetime, invert_logic, unique_id=None):
-        """Initialize the RPi binary sensor."""
-        self._attr_name = name or DEVICE_DEFAULT_NAME
-        self._attr_unique_id = unique_id
-        self._attr_should_poll = False
+    def __init__(self, hub, name, port, unique_id, active_low, bias, debounce):
+        _LOGGER.debug(f"GPIODBinarySensor init: {port} - {name} - {unique_id}")
+        self._hub = hub
+        self.name = name
+        self.unique_id = unique_id
         self._port = port
-        self._pull_mode = pull_mode
-        self._bouncetime = bouncetime
-        self._invert_logic = invert_logic
-        self._state = None
-
-        setup_input(self._port, self._pull_mode)
+        self._active_low = active_low
+        self._bias = bias
+        self._debounce = debounce
 
-        def edge_detected(port):
-            """Edge detection handler."""
-            if self.hass is not None:
-                self.hass.add_job(self.async_read_gpio)
-
-        edge_detect(self._port, edge_detected, self._bouncetime)
-
-    @property
-    def is_on(self):
-        """Return the state of the entity."""
-        return self._state != self._invert_logic
+    async def async_added_to_hass(self) -> None:
+        await super().async_added_to_hass()
+        self._hub.add_sensor(self, self._port, self._active_low, self._bias, self._debounce)
+        self.async_write_ha_state()
 
     def update(self):
-        """Update the GPIO state."""
-        self._state = read_input(self._port)
+        self.is_on = self._hub.update(self._port)
+        self.schedule_update_ha_state(False)
```

Comment:
```suggestion
```
This should cause update to happen twice, please verify if it is really needed

Reply:
I didn't notice something happens twice. @thecode, why do you suspect that? Maybe I'm missing something.

Reply:
You can't notice if something update twice, only if you enable debug logging for core (which is really hard to read later) or check the DB.
What you can do is remove this line, if everything is updated correctly (for example when you change the input state it updates in HA) it means it is not necessary, from looking at Home Assistant core code this should not be needed, please verify.